tensorboard/data/server/gcs/auth.rs

```diff
@@ -0,0 +1,300 @@
+//! ["Refreshing Access Tokens"]: https://www.oauth.com/oauth2-servers/access-tokens/refreshing-access-tokens/
+
+use log::{debug, info, warn};
+use serde::{Deserialize, Serialize};
+use std::fmt::{self, Debug};
+use std::fs::File;
+use std::io::BufReader;
+use std::path::PathBuf;
+use std::sync::RwLock;
+use std::time::{Duration, Instant};
+
+use reqwest::blocking::{Client as HttpClient, RequestBuilder};
+
+const OAUTH_REFRESH_TOKEN_ENDPOINT: &str = "https://www.googleapis.com/oauth2/v3/token";
+
+/// A set of refreshable OAuth credentials plus a potentially active token. Use
+/// [`authenticate`][Self::authenticate] to add an `Authorization` header to an outgoing request,
+/// fetching a fresh access token if necessary.
+///
+/// A `TokenStore` may be freely shared among threads; it synchronizes internally if needed.
+pub struct TokenStore {
+    creds: Credentials,
+    token: RwLock<Option<BoundedToken>>,
+}
+
+impl TokenStore {
+    /// Creates a new token store from the given credentials.
+    ///
+    /// This operation is cheap and does not actually fetch any access tokens.
+    pub fn new(creds: Credentials) -> Self {
+        Self {
+            creds,
+            token: RwLock::new(None),
+        }
+    }
+}
+
+/// An access token that's valid until a particular point in time.
+#[derive(Debug)]
+struct BoundedToken {
+    access_token: AccessToken,
+    expires: Instant,
+}
+
+impl BoundedToken {
+    /// Tests whether this token will still be valid at the given instant.
+    fn valid_at(&self, when: Instant) -> bool {
+        when < self.expires
+    }
+}
+
+/// Private access token module to prevent accidental leaking of tokens into logs, etc. An access
+/// token can be attached to a request, but cannot be directly extracted.
+mod access_token {
+    use super::*;
+    #[derive(Deserialize)]
+    pub struct AccessToken(String);
+    impl AccessToken {
+        /// Attaches this token to an outgoing request.
+        pub fn authenticate(&self, rb: RequestBuilder) -> RequestBuilder {
+            rb.bearer_auth(&self.0)
+        }
+    }
+    impl Debug for AccessToken {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_tuple("AccessToken")
+                .field(&format_args!("_"))
+                .finish()
+        }
+    }
+}
+use access_token::AccessToken;
+
+impl TokenStore {
+    /// Attempts to attach an access token to the given outgoing request.
+    ///
+    /// A cached token will be reused if it is expected to be valid for at least `lifetime`.
+    /// Otherwise, a new token will be fetched and stored.
+    pub fn authenticate(
+        &self,
+        rb: RequestBuilder,
+        http: &HttpClient,
+        lifetime: Duration,
+    ) -> RequestBuilder {
+        if self.creds.anonymous() {
+            return rb;
+        }
+        let token = self.token.read().expect("failed to read auth token");
+        if let Some(t) = Self::check_expiry(&*token, lifetime) {
```

Comment:
Thinking out loud:
At first glance, I was confused by what '&*' meant, but I think it makes sense now.  The type of `token` is a `RwLockReadGuard<Option<BoundedToken>>`, and what we really want to check is the `Option<BoundedToken>`.  Doing the deref '*' removes the guard, and the next '&' gets a reference to the Option, since we don't want to do a value copy.

If we simply did a `&token`, the code still compiles, but then we're just getting a reference to the guard, which is not the same as the reference to what it is guarding (the Option).